### PR TITLE
Simplify handle indexing

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105696
* __->__ #105006

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -504,6 +504,10 @@ def __init__(
         self._training_state = HandleTrainingState.IDLE
         self._debug_level = dist.get_debug_level()
         self._fully_sharded_module = fully_sharded_module
+        # Maps each handle to its index in `all_handles`, which must be the
+        # same across ranks for the execution order validation to work
+        self._handle_index = None
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:509]: nit: 
```suggestion
        self._handle_index: Optional[int] = None
```

**Review by awgu — APPROVED** [review]
This is great!

**Comment by voznesenskym** [comment]
squashed all of them here

File: torch/distributed/fsdp/_exec_order_utils.py [hunk 2]
```diff
@@ -85,8 +78,8 @@ def init(
         for handle in traversal_utils._get_fsdp_handles(root_module):
             index = len(self.all_handles)
             self.all_handles.append(handle)
-            self.handle_to_handle_index[handle] = index
-        self.param_to_fqn = _get_param_to_fqns(root_module)
+            handle._handle_index = index
+            self.param_to_fqn = _get_param_to_fqns(root_module)
```
Comment (awgu) [inline torch/distributed/fsdp/_exec_order_utils.py:82]: Should this be indented?

File: torch/distributed/fsdp/flat_param.py [hunk 3]
```diff
@@ -510,6 +510,19 @@ def __init__(
         # it points to parameterizes behavior. We use the following attribute
         # to track which tensor data the parameters are unsharded views into.
         self._unsharded_flat_param_for_skipped_views: Optional[Tensor] = None
+        # The index in the state's `all_handles`, which must be the
+        # same across ranks for the execution order validation to work
+        self._handle_index = None
+        # Index in handles_to_pre_forward_order
+        self._pre_forward_order_index = None
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:517]:
```suggestion
        self._pre_forward_order_index: Optional[int] = None
```

File: torch/distributed/fsdp/flat_param.py [hunk 4]
```diff
@@ -510,6 +510,19 @@ def __init__(
...
+        # Index in `handles_post_forward_order`
+        self._post_forward_index = None
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:519]:
```suggestion
        self._post_forward_index: Optional[int] = None
```

File: torch/distributed/fsdp/_exec_order_utils.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/_exec_order_utils.py:82]: @voznesenskym Can you check this (possibly in a follow-up)? I don't think this should be indented. It does not hurt correctness but seems like something to fix.

Comment (awgu) [reply torch/distributed/fsdp/_exec_order_utils.py:82]: I can fix this in https://github.com/pytorch/pytorch/pull/105738/.

**Pull request closed without merging**